presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,430 @@
+                }
+                return createJoinEnumerationResult(planNode);
+            }
+            return chooseJoinOrder(nodes, outputSymbols);
+        }
+
+        private static boolean isJoinEqualityCondition(Expression expression)
+        {
+            return expression instanceof ComparisonExpression
+                    && ((ComparisonExpression) expression).getOperator() == EQUAL
+                    && ((ComparisonExpression) expression).getLeft() instanceof SymbolReference
+                    && ((ComparisonExpression) expression).getRight() instanceof SymbolReference;
+        }
+
+        private static JoinNode.EquiJoinClause toEquiJoinClause(ComparisonExpression equality, Set<Symbol> leftSymbols)
+        {
+            Symbol leftSymbol = Symbol.from(equality.getLeft());
+            Symbol rightSymbol = Symbol.from(equality.getRight());
+            JoinNode.EquiJoinClause equiJoinClause = new JoinNode.EquiJoinClause(leftSymbol, rightSymbol);
+            return leftSymbols.contains(leftSymbol) ? equiJoinClause : equiJoinClause.flip();
+        }
+
+        private JoinEnumerationResult setJoinNodeProperties(JoinNode joinNode)
+        {
+            // TODO avoid stat (but not cost) recalculation for all considered (distribution,flip) pairs, since resulting relation is the same in all case
+            if (isAtMostScalar(joinNode.getRight(), lookup)) {
+                return createJoinEnumerationResult(joinNode.withDistributionType(REPLICATED));
+            }
+            if (isAtMostScalar(joinNode.getLeft(), lookup)) {
+                return createJoinEnumerationResult(joinNode.flipChildren().withDistributionType(REPLICATED));
+            }
+
+            List<JoinEnumerationResult> possibleJoinNodes = new ArrayList<>();
+            FeaturesConfig.JoinDistributionType joinDistributionType = getJoinDistributionType(session);
+            if (joinDistributionType.canRepartition() && !joinNode.isCrossJoin()) {
+                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.withDistributionType(PARTITIONED)));
+                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.flipChildren().withDistributionType(PARTITIONED)));
+            }
+            if (joinDistributionType.canReplicate()) {
+                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.withDistributionType(REPLICATED)));
+                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.flipChildren().withDistributionType(REPLICATED)));
+            }
+            if (possibleJoinNodes.stream().anyMatch(UNKNOWN_COST_RESULT::equals)) {
+                return UNKNOWN_COST_RESULT;
+            }
+            return resultComparator.min(possibleJoinNodes);
+        }
+
+        private JoinEnumerationResult createJoinEnumerationResult(PlanNode planNode)
+        {
+            return JoinEnumerationResult.createJoinEnumerationResult(Optional.of(planNode), costProvider.getCumulativeCost(planNode));
```

Comment:
Inline `JoinEnumerationResult.createJoinEnumerationResult`

Reply:
haha https://github.com/prestodb/presto/pull/10753/files/0aa2047cb77f768eeb5fc42acab9348ea2ee85ea..904bd5ed9685e7b769aa0142e08abb97c962349e#r198940384.

To me it's pretty equivalent the extra bit of indirection vs. the slightly clunkier `setJoinNodeProperties`, and I don't really feel like changing it again, so leaving as is.

Reply:
Oh, sorry. I overlooked the second parameter. Yes, sure, leave it as is.